pkg/bulk/engine.go

```
@@ -0,0 +1,221 @@
+		return fmt.Errorf("failed to parse YAML from %s: %w", configPath, err)
+	}
+
+	timeout, err := time.ParseDuration(opts.Timeout)
+	if err != nil {
+		return fmt.Errorf("invalid timeout duration: %w", err)
+	}
+
+	log.Debug("Discovering all tags to calculate total job count...")
+	type job struct {
+		spec *ImageSpec
+		tag  string
+	}
+	var jobsToRun []job
+	var discoveryErrors *multierror.Error
+
+	for i := range config.Images {
+		imageSpec := &config.Images[i]
+		tagsToPatch, err := FindTagsToPatch(imageSpec)
+		if err != nil {
+			discoveryErrors = multierror.Append(discoveryErrors, fmt.Errorf("error discovering tags for '%s': %w", imageSpec.Name, err))
+			continue
+		}
+		for _, tag := range tagsToPatch {
+			jobsToRun = append(jobsToRun, job{spec: imageSpec, tag: tag})
+		}
+	}
+
+	if discoveryErrors.ErrorOrNil() != nil {
+		log.Warnf("Encountered errors during tag discovery phase:\n%s", discoveryErrors.Error())
+	}
+
+	if len(jobsToRun) == 0 {
+		log.Warn("No tags found to patch across all image specs.")
+		return nil
+	}
+
+	log.Debugf("Total number of patch jobs to execute: %d", len(jobsToRun))
+
+	numWorkers := runtime.NumCPU()
+	log.Debugf("initializing worker pool with %d concurrent workers.", numWorkers)
+
+	var wg sync.WaitGroup
+	var mu sync.Mutex
+
+	jobsChan := make(chan job, len(jobsToRun))
+	errChan := make(chan error, len(jobsToRun))
+	results := make([]patchJobStatus, 0, len(jobsToRun))
+
+	log.Infof("Starting bulk patch for %d image(s) defined in %s...", len(config.Images), configPath)
+
+	for w := 1; w <= numWorkers; w++ {
+		wg.Add(1)
+		go func(workerID int) {
+			defer wg.Done()
+
+			for j := range jobsChan {
+				spec := j.spec
+				tag := j.tag
+				imageWithTag := fmt.Sprintf("%s:%s", spec.Image, tag)
+				targetTag, err := resolveTargetTag(spec.Target, tag)
+				if err != nil {
+					errChan <- fmt.Errorf("worker %d: error resolving target tag for '%s:%s': %w", workerID, spec.Name, tag, err)
```

I'm wondering if we should implement ignore errors similar to the multiplatform work, i.e. give the option to continue with other tasks if one fails?